Add support for scaling up

rmetzger opened the pull request:
Based on and includes https://github.com/tillrohrmann/flink/pull/16.

This is really a minimal implementation for scaling up the job on additional slots.
For production use, I guess a configurable "cooldown" phase, conditions for scaling (at least n more slots, doubling of slots, ...) and more is needed.

rmetzger committed:
[FLINK-XXXX] Add support for scaling up

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/allocator/SlotAllocator.java:
```diff
@@ -30,12 +31,12 @@
      *     could be run with the given slots
      */
     Optional<T> determineParallelism(
-            JobInformation jobInformation, Collection<SlotInfoWithUtilization> freeSlots);
```
We will need the utilization at some point once we want to distribute tasks in specific ways (on as few task executors as possible / evenly); what is your goal here?

rmetzger replied:
I want to call `determineParallelism` with ALL slots, instead of the free slots. I believe we should not distribute increments of new slots, but always the total number of available slots.
```java
slotAllocator.determineParallelism(
                                new JobGraphJobInformation(jobGraph),
                                declarativeSlotPool.getAllSlotsInformation());
```

Maybe we need to revisit the DeclarativeSlotPool.getAllSlotsInformation() method? I guess we could also calculate the taskExecutorUtilization for allocated slots as well?

zentol replied:
Yes, we could look into changing this method, but ideally we get by without having to modify the SlotPool interface in significant ways.
Alternatively we could always compute it within the scheduler on the fly. That may be a better option than forcing this computation upon all usages of this method.

rmetzger replied:
I don't see an easy way to solve this:

I'm not sure if computing it in the scheduler is the right separation of concerns. For calculating the utilization, I need to know whether the slot is free or allocated. That's something that the scheduler can not get from the `SlotInfo` interface.

What's the problem with modifying the SlotPool interface? (Is changing the return type of a method significant?)

zentol replied:
You would then need to touch the SlotpoolImpl which would be a waste of time, both for developers and in terms of performance (unless you go with a broken implementation that always returns 0, but, you know...).

Computing it one the fly isn't that difficult; you get the free slots, extract utilization by task manager, and then query it for all slots returned by getAllSlotsInformation.

Nevertheless, I think the best short-term solution is to stick with plain `SlotInfos`. There's no immediate need for the utilization, and whether we solve it now or later probably won't change anything.

rmetzger replied:
Allright, then let's change this once we need the utilization in the new scheduler.

zentol commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerClusterITCase.java:
```diff
@@ -94,6 +95,38 @@ public void testAutomaticScaleDownInCaseOfLostSlots() throws InterruptedExceptio
+        OnceBlockingNoOpInvokable.waitUntilOpsAreRunning();
+
+        log.info("Start additional TaskManager to scale up to the full parallelism.");
+        miniCluster.startTaskManager();
+        OnceBlockingNoOpInvokable.resetInstanceCount(); // we expect a restart
```
these 2 lines should be switched in case the restart is really quick.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerNG.java:
```diff
@@ -902,6 +912,39 @@ private void deploySafely(ExecutionVertex executionVertex) {
+
+                if (potentialNewParallelism.isPresent()
+                        && isParallelismHigher(potentialNewParallelism.get())) {
+                    transitionToState(new Restarting(executionGraph, 0L));
```
Something should be logged along the lines of "restarting the job to scale up". It could also be interesting to log the cumulative different between the old and new parallelism.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerNG.java:
```diff
@@ -114,19 +115,22 @@
 
     private static final Logger LOG = LoggerFactory.getLogger(DeclarativeSchedulerNG.class);
 
+    // note: do not mutate this jobGraph!
```
A less error-prone alternative might be to compute a constant `JobGraphJobInformation` from the initial JobGraph. We need few enough bits of information that this shouldn't be a problem.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerNG.java:
```diff
@@ -640,7 +645,11 @@ private void runIfState(State expectedState, Runnable action, long delay, TimeUn
+                        try {
+                            action.run();
+                        } catch (Throwable t) {
+                            LOG.warn("Error in runnable", t);
```
Chances are that this should go straight to the FatalExitExceptionHandler#INSTANCE.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerNG.java:
```diff
@@ -902,6 +912,39 @@ private void deploySafely(ExecutionVertex executionVertex) {
+            int availableSlots = declarativeSlotPool.getFreeSlotsInformation().size();
+
+            if (availableSlots > 0) {
+                LOG.info(
```
log as debug or maybe even trace; the period check shouldn't be too interesting.

rmetzger commented:
Good feedback, thanks! I'll address it soon.

zentol commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerNG.java:
```diff
@@ -170,7 +168,7 @@
     private State state = new Created();
 
     public DeclarativeSchedulerNG(
-            JobGraph jobGraph,
+            JobGraph jobInformation,
```
search&replace was a bit too eager ;)

rmetzger replied:
ah, that was the argument the refactoring tool asked me about. Should have looked closer 🤦

rmetzger commented:
@zentol I should have addressed all your comments. Thanks for your help!

@tillrohrmann this is ready for a final review, or merge.

rmetzger commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/JobGraphJobInformation.java:
```diff
@@ -40,11 +48,47 @@ public JobGraphJobInformation(JobGraph jobGraph) {
 
     @Override
-    public VertexInformation getVertexInformation(JobVertexID jobVertexId) {
+    public JobInformation.VertexInformation getVertexInformation(JobVertexID jobVertexId) {
```
maybe I need to change my IntelliJ inspections ;)

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java:
```diff
@@ -742,13 +746,24 @@ public void goToFailing(
+    }
+
+    @Override
+    public Optional<? extends VertexAssignment> getVertexAssignmentWithAllSlots() {
```
Wildcards in return types are discouraged because they only make the API less comfortable. Please return `Optional<VertexAssignment>`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeScheduler.java:
```diff
@@ -777,7 +792,12 @@ public void runIfState(State expectedState, Runnable action, Duration delay) {
+                        try {
+                            action.run();
+                        } catch (Throwable t) {
+                            FatalExitExceptionHandler.INSTANCE.uncaughtException(
+                                    Thread.currentThread(), t);
```
Let's keep in mind to pass in a proper `FatalExceptionHandler` to the `DeclarativeScheduler`. Maybe add a TODO for that.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/Executing.java:
```diff
@@ -129,6 +134,43 @@ private void handleDeploymentFailure(ExecutionVertex executionVertex, JobExcepti
 
+    @Override
+    public void notifyNewResourcesAvailable() {
+        int availableSlots = context.getAvailableSlots();
+
+        if (availableSlots > 0) {
+            // check if rescaling would increase the parallelism of at least one vertex
+
+            final Optional<? extends VertexAssignment> potentialNewParallelism =
+                    context.getVertexAssignmentWithAllSlots();
```
Instead of introducing two new methods one could introduce a method `boolean Context.canScaleUp()` which is implemented by the `DeclarativeScheduler`. The implementation could then call into a `RescalingStrategy` which makes this decision. That way the interaction between the `Executing` state and the `Executing.Context` implementation is a bit easier.

rmetzger replied:
Thx for the feedback. I moved the logic into the scheduler.

I assume introducing the RescalingStrategy should be done once we've agreed on an interface in the design document, and then done in a separate PR.

tillrohrmann replied:
Yes, exactly.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/Executing.java:
```diff
@@ -184,6 +226,10 @@ void goToFailing(
                 OperatorCoordinatorHandler operatorCoordinatorHandler,
                 Throwable failureCause);
+
+        int getAvailableSlots();
+
+        Optional<? extends VertexAssignment> getVertexAssignmentWithAllSlots();
```
I think we can simplify this a bit by introducing a `boolean canScaleUp` method.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/JobGraphJobInformation.java:
```diff
@@ -40,11 +48,47 @@ public JobGraphJobInformation(JobGraph jobGraph) {
+    }
+
+    /** Returns a copy of a jobGraph that can be mutated. */
+    public JobGraph getMutableJobGraph() throws IOException, ClassNotFoundException {
```
I think this method name is a bit misleading. The JavaDoc say that it copies the `JobGraph`. Why not calling this method then `copyJobGraph()`?

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerClusterITCase.java:
```diff
@@ -94,6 +95,37 @@ public void testAutomaticScaleDownInCaseOfLostSlots() throws InterruptedExceptio
+        int targetInstanceCount = NUMBER_SLOTS_PER_TASK_MANAGER * (NUMBER_TASK_MANAGERS + 1);
+        final JobGraph jobGraph = createBlockingJobGraph(targetInstanceCount);
+
+        // initially only expect 1 TM
```
Comment is wrong. One expects `NUMBER_TASK_MANAGERS`.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/DeclarativeSchedulerClusterITCase.java:
```diff
@@ -94,6 +95,37 @@ public void testAutomaticScaleDownInCaseOfLostSlots() throws InterruptedExceptio
+        miniCluster.startTaskManager();
+
+        log.info("Waiting until Invokable is running with higher parallelism");
+        while (OnceBlockingNoOpInvokable.getInstanceCount() < targetInstanceCount) {
+            Thread.sleep(50);
+        }
```
Can we do this w/o busy loop waiting? I think there is very rarely the need for this pattern.

tillrohrmann reviewed (changes requested):
Thanks for creating this PR @rmetzger. The change looks conceptually good. I had a couple of minor comments. Please take a look.

rmetzger commented:
Thanks for your feedback! 
I addressed all of your comments. I assume introducing the `RescalingStrategy` is a separate item, that needs a design review first.

tillrohrmann reviewed (changes requested):
Thanks for updating this PR @rmetzger. The change looks good now. Only spotless does not pass currently.

rmetzger commented:
Damn, I'm getting too used to the magic reformatter :) I pushed a version that at least passes the compile stage now.

tillrohrmann reviewed (approved):
Looks good to me. Merging once AZP gives green light.

tillrohrmann commented:
Merged via 3eaaae1839e868d2e629a6bffbafc4f2be76b040
